lib/iris/common/resolve.py

```diff
@@ -0,0 +1,1541 @@
+# Copyright Iris contributors
+#
+# This file is part of Iris and is released under the LGPL license.
+# See COPYING and COPYING.LESSER in the root of the repository for full
+# licensing details.
+
+from collections import namedtuple
+from collections.abc import Iterable
+import logging
+
+from dask.array.core import broadcast_shapes
+
+from iris.common import LENIENT
+
+
+__all__ = ["Resolve"]
```

Comment:
@lbdreyer For me, this class doesn't belong in `iris.common.metadata`. 

It's certainly using `iris.common.metadata` but it's primarily dealing with higher order objects, such as cubes, coordinates and factories. So I believe it stands in its own right.

Naming things can be the hardest part, so I'm open to suggestions, but I'd be keen to steer clear of incorporating the word `metadata` in the module or class name... as a term I think it's already too overloaded within `iris`. Hmmm, I'll have a think...

Also, as I said previously, I've not finished fully commenting the code or adding doc-strings, so that will come.